### PR TITLE
Accelerator Refactor/RPC + Sharded

### PR DESCRIPTION
**Pull request by justusschock**
This should be the last PR adding new classes for the accelerator refactor (#5616 ).

It should be merged after #5719 .

It adds support for sharded training and RPC/Model parallelism.

### CODE REVIEW

**Review by tchaton — APPROVED** [review]
LGTM ! I think we will have to be careful with duplicated code for Plugin as Shared.

File: pytorch_lightning/plugins/training_type/rpc.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        self,
+        parallel_devices,
+        num_nodes=1,
+        cluster_environment: ClusterEnvironment = None,
+        sync_batchnorm=False,
+        rpc_timeout_sec: float = DEFAULT_RPC_TIMEOUT_SEC,
+        **kwargs
+    ):
+        self.rpc_timeout_sec = rpc_timeout_sec
+        self.rpc_initialized = False
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/rpc.py:51]:
```suggestion
        self._is_rpc_initialized = False
```
also as protected as it shall not be changed by user

File: pytorch_lightning/plugins/training_type/rpc.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        Override to decide whether to skip train/test function after shutdown completed.
+        Usually RPC shutdown is a join/exit function, afterwards we want to exit the process.
+        Returns: Whether to return after rpc exit.
+        """
+        raise NotImplementedError
+
+    def worker_optimizer_step(self, model: LightningModule, opt_idx: int, *args, **kwargs) -> None:
+        """
+        Called when optimizer step is run on the main process. Used to signal any RPC workers to run optimizer step.
+        Args:
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/rpc.py:111]:
```suggestion

        Args:
```

File: pytorch_lightning/plugins/training_type/rpc.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        Called when main rpc connection has been established.
+        Args:
+            trainer: The trainer object.
+        """
+        raise NotImplementedError
+
+    def on_accelerator_exit_rpc_process(self) -> None:
+        """
+        Called to exit RPC process within the accelerator, that is being managed by main process.
+        Args:
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/rpc.py:89]:
```suggestion

        Args:
```

File: pytorch_lightning/plugins/training_type/rpc.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+            last_filepath: The filepath to save the model to.
+            trainer: The trainer object.
+            pl_module: The LightningModule.
+        """
+        raise NotImplementedError
+
+    def on_main_rpc_connection(self, trainer) -> None:
+        """
+        Called when main rpc connection has been established.
+        Args:
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/rpc.py:81]:
```suggestion

        Args:
```

File: pytorch_lightning/plugins/training_type/rpc.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        os.environ['MASTER_PORT'] = os.getenv('RPC_MASTER_PORT', '15000')
+        rpc.init_rpc(f"worker{global_rank}", rank=global_rank, world_size=world_size)
+        rpc._set_rpc_timeout(self.rpc_timeout_sec)
+        self.rpc_initialized = True
+
+    def rpc_save_model(self, save_model_fn, last_filepath, trainer, pl_module) -> None:
+        """
+        Override to save model to disk.
+        This is required as the main process will be required to handle aggregating model states from RPC processes.
+        Args:
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/rpc.py:70]:
```suggestion

        Args:
```

File: pytorch_lightning/plugins/training_type/rpc.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        if self.rpc_initialized:
+            torch.distributed.rpc.shutdown()
+            self.rpc_initialized = False
+
+    @property
+    def return_after_exit_rpc_process(self) -> bool:
+        """
+        Override to decide whether to skip train/test function after shutdown completed.
+        Usually RPC shutdown is a join/exit function, afterwards we want to exit the process.
+        Returns: Whether to return after rpc exit.
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/rpc.py:104]:
```suggestion

        Returns:
            Whether to return after RPC exit.
```

File: pytorch_lightning/plugins/training_type/rpc_sequential.py [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+log = logging.getLogger(__name__)
+
+if FAIRSCALE_PIPE_AVAILABLE:
+    import fairscale.nn.model_parallel as mpu
+    from fairscale.nn import PipeRPCWrapper
+    from fairscale.nn.pipe import balance as pipe_balance
+    from fairscale.nn.pipe import rpc as rpc_pipe
+    from fairscale.nn.pipe.pipeline import PipelineStyle
+
+
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/rpc_sequential.py:40]:
```suggestion
if FAIRSCALE_PIPE_AVAILABLE:
    import fairscale.nn.model_parallel as mpu
    from fairscale.nn import PipeRPCWrapper
    from fairscale.nn.pipe import balance as pipe_balance
    from fairscale.nn.pipe import rpc as rpc_pipe
    from fairscale.nn.pipe.pipeline import PipelineStyle

log = logging.getLogger(__name__)


```

File: pytorch_lightning/plugins/training_type/rpc_sequential.py [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+            model.sequential_module.configure_optimizers = model.configure_optimizers
+
+            # Update references for main process to access correct lightning functions when calling RPC
+            model.sequential_module.module.model.trainer = model.trainer
+            model.sequential_module.module.model.configure_optimizers = model.configure_optimizers
+
+        else:
+            raise MisconfigurationException(
+                'Could not find a PipeLightningModule within the model. '
+                'Did you defined set your sequential model as an `sequential_module` attribute of your model ?'
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/rpc_sequential.py:203]:
```suggestion
                'Did you defined set your sequential model as a `sequential_module` attribute of your model?'
```

File: pytorch_lightning/plugins/training_type/sharded_spawn.py [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+                optim_class = type(optimizer)
+                zero_optimizer = OSS(params=optimizer.param_groups, optim=optim_class, **optimizer.defaults)
+                optimizers[x] = zero_optimizer
+                del optimizer
+        trainer = self.lightning_module.trainer
+        trainer.optimizers = trainer.convert_to_lightning_optimizers(optimizers)
+
+    def _wrap_optimizers(self):
+        trainer = self.model.trainer
+        if trainer.testing is True:
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/sharded_spawn.py:36]:
```suggestion
        if trainer.testing:
```

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
just add chlog :]

**Review by awaelchli — APPROVED** [review]

**Pull request merged**